config/config.go

```diff
@@ -1055,6 +1055,20 @@ func (cfg *ConsensusConfig) Propose(round int32) time.Duration {
 	) * time.Nanosecond
 }
 
+// ProposeWithCustomTimeout is identical to Propose. However,
+// it calculates the amount of time to wait for a proposal using the supplied
+// customTimeout.
+// If customTimeout is 0, the default TimeoutCommit is used.
+func (cfg *ConsensusConfig) ProposeWithCustomTimeout(round int32, customTimeout time.Duration) time.Duration {
+	// this is to capture any unforeseen cases where the customTimeout is 0
+	var timeoutPropose time.Duration
+	if customTimeout == 0 {
+		// falling back to default timeout
+		timeoutPropose = cfg.TimeoutPropose
+	}
+	return time.Duration(timeoutPropose.Nanoseconds()+cfg.TimeoutProposeDelta.Nanoseconds()*int64(round)) * time.Nanosecond
+}
```

Comment:
[blocking]

doesn't this always propose with 0s? are we not using the custom timeout that we pass here for anything if its not 0?


> ProposeWithCustomTimeout is identical to Propose. However,
// it calculates the amount of time to wait for a proposal using the supplied
// customTimeout.

is this supposed to get different results from `Propose`? when I test this locally and pass the same configured timeouts it gets zero when `Propose` gets 3s

am I missing something?

Reply:
Thanks for pointing that out! I was simplifying the if-else statement during cleanup, and the timeoutPropose should have been initialized with the custom Timeout. It's fixed now, and I've added a test as well.

Reply:
Please see https://github.com/celestiaorg/celestia-core/pull/1494/commits/10be8b14bcb846d164abdd2b68d3102b7e1aa110

Reply:
I ran an e2e test on the app side and can confirm that timeouts are properly set and are based on the app version (and non-zero values):
```
make test-e2e MajorUpgradeToV3
--> Running end to end tests
go run ./test/e2e MajorUpgradeToV3
test-e2e2024/10/11 14:30:20 === RUN MajorUpgradeToV3
test-e2e2024/10/11 14:30:20 Creating testnet
INFO[2024-10-11T14:30:20-07:00]log/logger.go:41 Current log level                             env_log_level=LOG_LEVEL log_level=info
{"level":"info","scope":"majorupgradetov3-20241011-143020","TestName":"MajorUpgradeToV3","time":"2024-10-11T14:30:43-07:00","message":"Knuu initialized"}
test-e2e2024/10/11 14:30:43 Running major upgrade to v3 test version pr-3882
test-e2e2024/10/11 14:30:43 Creating genesis nodes
test-e2e2024/10/11 14:30:44 Creating txsim
{"level":"info","name":"txsim","image":"ghcr.io/celestiaorg/txsim:pr-3882","time":"2024-10-11T14:30:48-07:00","message":"setting image for tx client"}
{"level":"info","name":"txsim","image":"ghcr.io/celestiaorg/txsim:pr-3882","args":"--key-path /var/folders/gl/8f7g64t91rxc6hsz_gj0mqfh0000gn/T/txsim --grpc-endpoint 10.37.191.24:9090 --poll-time 1s --seed 42 --blob 1 --blob-amounts 100 --blob-sizes 100-2000 --upgrade-schedule 4:3 --blob-share-version 0","time":"2024-10-11T14:30:48-07:00","message":"created tx client"}
test-e2e2024/10/11 14:30:48 Setting up testnet
{"level":"info","name":"val0","directory":"/var/folders/gl/8f7g64t91rxc6hsz_gj0mqfh0000gn/T/val0","time":"2024-10-11T14:30:48-07:00","message":"Creating validator's config and data directories"}
{"level":"info","name":"val1","directory":"/var/folders/gl/8f7g64t91rxc6hsz_gj0mqfh0000gn/T/val1","time":"2024-10-11T14:30:48-07:00","message":"Creating validator's config and data directories"}
{"level":"info","name":"val2","directory":"/var/folders/gl/8f7g64t91rxc6hsz_gj0mqfh0000gn/T/val2","time":"2024-10-11T14:30:48-07:00","message":"Creating validator's config and data directories"}
{"level":"info","name":"val3","directory":"/var/folders/gl/8f7g64t91rxc6hsz_gj0mqfh0000gn/T/val3","time":"2024-10-11T14:30:48-07:00","message":"Creating validator's config and data directories"}
test-e2e2024/10/11 14:30:48 Starting testnet
{"level":"info","time":"2024-10-11T14:30:54-07:00","message":"create endpoint proxies for genesis nodes"}
{"level":"info","name":"val0","version":"pr-3882","time":"2024-10-11T14:31:11-07:00","message":"started and ports forwarded"}
{"level":"info","name":"val1","version":"pr-3882","time":"2024-10-11T14:31:14-07:00","message":"started and ports forwarded"}
{"level":"info","name":"val2","version":"pr-3882","time":"2024-10-11T14:31:22-07:00","message":"started and ports forwarded"}
{"level":"info","name":"val3","version":"pr-3882","time":"2024-10-11T14:31:27-07:00","message":"started and ports forwarded"}
{"level":"info","time":"2024-10-11T14:31:27-07:00","message":"waiting for genesis nodes to sync"}
{"level":"info","name":"val0","time":"2024-10-11T14:31:27-07:00","message":"waiting for node to sync"}
{"level":"debug","RPC Address":"http://151.115.14.45:80/val0-26657","time":"2024-10-11T14:31:27-07:00","message":"Creating HTTP client for node"}
{"level":"info","name":"val0","attempt":0,"time":"2024-10-11T14:31:27-07:00","message":"node is not synced yet, waiting..."}
{"level":"info","name":"val0","attempt":1,"time":"2024-10-11T14:31:27-07:00","message":"node is not synced yet, waiting..."}
{"level":"info","name":"val0","attempt":2,"time":"2024-10-11T14:31:28-07:00","message":"node is not synced yet, waiting..."}
{"level":"info","name":"val0","attempt":3,"time":"2024-10-11T14:31:31-07:00","message":"node is not synced yet, waiting..."}
{"level":"info","name":"val0","attempt":4,"time":"2024-10-11T14:31:34-07:00","message":"node is not synced yet, waiting..."}
{"level":"info","name":"val0","attempt":5,"time":"2024-10-11T14:31:38-07:00","message":"node is not synced yet, waiting..."}
{"level":"info","attempts":6,"name":"val0","time":"2024-10-11T14:31:43-07:00","message":"node has synced"}
{"level":"info","name":"val1","time":"2024-10-11T14:31:43-07:00","message":"waiting for node to sync"}
{"level":"debug","RPC Address":"http://151.115.14.45:80/val1-26657","time":"2024-10-11T14:31:43-07:00","message":"Creating HTTP client for node"}
{"level":"info","attempts":0,"name":"val1","time":"2024-10-11T14:31:43-07:00","message":"node has synced"}
{"level":"info","name":"val2","time":"2024-10-11T14:31:43-07:00","message":"waiting for node to sync"}
{"level":"debug","RPC Address":"http://151.115.14.45:80/val2-26657","time":"2024-10-11T14:31:43-07:00","message":"Creating HTTP client for node"}
{"level":"info","attempts":0,"name":"val2","time":"2024-10-11T14:31:44-07:00","message":"node has synced"}
{"level":"info","name":"val3","time":"2024-10-11T14:31:44-07:00","message":"waiting for node to sync"}
{"level":"debug","RPC Address":"http://151.115.14.45:80/val3-26657","time":"2024-10-11T14:31:44-07:00","message":"Creating HTTP client for node"}
{"level":"info","attempts":0,"name":"val3","time":"2024-10-11T14:31:44-07:00","message":"node has synced"}
{"level":"info","name":"txsim","time":"2024-10-11T14:31:45-07:00","message":"txsim started"}
test-e2e2024/10/11 14:32:04 checking initial timeouts
{"level":"debug","RPC Address":"http://151.115.14.45:80/val0-26657","time":"2024-10-11T14:32:04-07:00","message":"Creating HTTP client for node"}
test-e2e2024/10/11 14:32:05 timeout commit: 11s, timeout propose: 10s
{"level":"debug","RPC Address":"http://151.115.14.45:80/val1-26657","time":"2024-10-11T14:32:05-07:00","message":"Creating HTTP client for node"}
test-e2e2024/10/11 14:32:05 timeout commit: 11s, timeout propose: 10s
{"level":"debug","RPC Address":"http://151.115.14.45:80/val2-26657","time":"2024-10-11T14:32:05-07:00","message":"Creating HTTP client for node"}
test-e2e2024/10/11 14:32:05 timeout commit: 11s, timeout propose: 10s
{"level":"debug","RPC Address":"http://151.115.14.45:80/val3-26657","time":"2024-10-11T14:32:05-07:00","message":"Creating HTTP client for node"}
test-e2e2024/10/11 14:32:06 timeout commit: 11s, timeout propose: 10s
test-e2e2024/10/11 14:32:06 waiting for upgrade
{"level":"debug","RPC Address":"http://151.115.14.45:80/val0-26657","time":"2024-10-11T14:32:06-07:00","message":"Creating HTTP client for node"}
test-e2e2024/10/11 14:32:35 height 5
test-e2e2024/10/11 14:32:35 timeout commit: 11s, timeout propose: 10s
test-e2e2024/10/11 14:33:05 height 8
test-e2e2024/10/11 14:33:05 timeout commit: 11s, timeout propose: 10s
test-e2e2024/10/11 14:33:35 height 11
test-e2e2024/10/11 14:33:35 timeout commit: 11s, timeout propose: 10s
test-e2e2024/10/11 14:34:05 height 13
test-e2e2024/10/11 14:34:05 timeout commit: 11s, timeout propose: 10s
test-e2e2024/10/11 14:34:35 height 16
test-e2e2024/10/11 14:34:35 timeout commit: 11s, timeout propose: 10s
test-e2e2024/10/11 14:35:05 height 20
test-e2e2024/10/11 14:35:05 timeout commit: 1s, timeout propose: 3s
{"level":"info","name":"val0","time":"2024-10-11T14:35:05-07:00","message":"upgrade is completed"}
{"level":"debug","RPC Address":"http://151.115.14.45:80/val1-26657","time":"2024-10-11T14:35:05-07:00","message":"Creating HTTP client for node"}
test-e2e2024/10/11 14:35:35 height 42
test-e2e2024/10/11 14:35:35 timeout commit: 1s, timeout propose: 3s
{"level":"info","name":"val1","time":"2024-10-11T14:35:35-07:00","message":"upgrade is completed"}
{"level":"debug","RPC Address":"http://151.115.14.45:80/val2-26657","time":"2024-10-11T14:35:35-07:00","message":"Creating HTTP client for node"}
test-e2e2024/10/11 14:36:05 height 64
test-e2e2024/10/11 14:36:05 timeout commit: 1s, timeout propose: 3s
{"level":"info","name":"val2","time":"2024-10-11T14:36:05-07:00","message":"upgrade is completed"}
{"level":"debug","RPC Address":"http://151.115.14.45:80/val3-26657","time":"2024-10-11T14:36:05-07:00","message":"Creating HTTP client for node"}
test-e2e2024/10/11 14:36:35 height 87
test-e2e2024/10/11 14:36:35 timeout commit: 1s, timeout propose: 3s
{"level":"info","name":"val3","time":"2024-10-11T14:36:35-07:00","message":"upgrade is completed"}
{"level":"info","upgradedHeight":20,"time":"2024-10-11T14:36:35-07:00","message":"checking timeouts"}
{"level":"debug","RPC Address":"http://151.115.14.45:80/val0-26657","time":"2024-10-11T14:36:35-07:00","message":"Creating HTTP client for node"}
{"level":"info","name":"val0","time":"2024-10-11T14:36:35-07:00","message":"checking timeouts"}
{"level":"info","name":"val0","time":"2024-10-11T14:36:45-07:00","message":"timeouts are checked"}
{"level":"debug","RPC Address":"http://151.115.14.45:80/val1-26657","time":"2024-10-11T14:36:45-07:00","message":"Creating HTTP client for node"}
{"level":"info","name":"val1","time":"2024-10-11T14:36:45-07:00","message":"checking timeouts"}
{"level":"info","name":"val1","time":"2024-10-11T14:36:55-07:00","message":"timeouts are checked"}
{"level":"debug","RPC Address":"http://151.115.14.45:80/val2-26657","time":"2024-10-11T14:36:55-07:00","message":"Creating HTTP client for node"}
{"level":"info","name":"val2","time":"2024-10-11T14:36:55-07:00","message":"checking timeouts"}
{"level":"info","name":"val2","time":"2024-10-11T14:37:05-07:00","message":"timeouts are checked"}
{"level":"debug","RPC Address":"http://151.115.14.45:80/val3-26657","time":"2024-10-11T14:37:05-07:00","message":"Creating HTTP client for node"}
{"level":"info","name":"val3","time":"2024-10-11T14:37:05-07:00","message":"checking timeouts"}
{"level":"info","name":"val3","time":"2024-10-11T14:37:14-07:00","message":"timeouts are checked"}
test-e2e2024/10/11 14:37:15 --- ✅ PASS: MajorUpgradeToV3 
```

Reply:
awesome, thanks!

at some point, we might want to add some simple unit tests for these functions to ensure that easy to miss stuff like this doesn't make it through

Reply:
Sure, this function was extended with the fallback option just yesterday, so I had initially missed adding a unit test. I’ve now added it in [this commit](https://github.com/celestiaorg/celestia-core/pull/1494/commits/363058a623c98f80c7c9d4b5241862200d5fb4b1). I've aimed to cover as much as possible in terms of unit testing in this PR. Please let me know if you find any missing tests.